### PR TITLE
Add category system to plugin configuration UI. Fixes #2411

### LINKED ISSUE #2411: Plugin categories
For example Combat / Skilling / QoL / Other.
The plugin list is getting HUGE 👍

### PR DESCRIPTION
**Pull request by AdamBallinger**
001447b626d2d6324d0fbc55c2c883f5
Adds category field to PluginDescriptor annotation.
New categories can be added in the PluginCategory enum.
New categories are automatically added to the config UI.

### CODE REVIEW

**Comment by deathbeam** [comment]
Well, I must say, this looks horrible. Maybe a dropdown instead?

**Comment by deathbeam** [comment]
Also, both utility and other should not be categories, that is basically uncategorized stuff that you dumped somewhere, so if you want to properly categorize them, remove these 2 and figure out something better.

**Comment by AdamBallinger** [comment]
Thanks for the feedback! Will get started on improvements.

**Comment by AdamBallinger** [comment]
Removed other category and replaced it with client and updated plugins accordingly. I think utility as a category is fine as plugins that use the utility category are exactly that; utility.
001447b626d2d6324d0fbc55c2c883f5

**Comment by deathbeam** [comment]
It still feels like having everything that do not fit anywhere in utility and client. I would add Items categories, then Minigames, Bosses, Indicators, Combat looks fine I guess.

**Comment by AdamBallinger** [comment]
Added 3 new categories and I feel like what remains in Utility / Client are hard to really categorise as anything else.

File: runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollPlugin.java [hunk 1]
```diff
@@ -87,7 +87,7 @@
 
 @PluginDescriptor(
 	name = "Clue Scroll",
-	category = PluginCategory.UTILITY
+	category = PluginCategory.ACTIVITY
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/cluescrolls/ClueScrollPlugin.java]: Clue scrolls are minigame

File: runelite-client/src/main/java/net/runelite/client/plugins/implings/ImplingsPlugin.java [hunk 2]
```diff
@@ -49,7 +49,7 @@
  */
 @PluginDescriptor(
 	name = "Implings",
-	category = PluginCategory.UTILITY
+	category = PluginCategory.ACTIVITY
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/implings/ImplingsPlugin.java]: Implings are minigame

File: runelite-client/src/main/java/net/runelite/client/plugins/kourendlibrary/KourendLibraryPlugin.java [hunk 3]
```diff
@@ -52,7 +52,7 @@
 
 @PluginDescriptor(
 	name = "Kourend Library",
-	category = PluginCategory.UTILITY
+	category = PluginCategory.ACTIVITY
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/kourendlibrary/KourendLibraryPlugin.java]: Isn't this minigame too? Sort of

File: runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java [hunk 4]
```diff
@@ -149,11 +151,21 @@ final void rebuildPluginList()
...
 		Map<String, JPanel> newChildren = new TreeMap<>();
 
+		children.clear();
+
 		pluginManager.getPlugins().stream()
 				.filter(plugin -> !plugin.getClass().getAnnotation(PluginDescriptor.class).hidden())
 				.sorted(Comparator.comparing(left -> left.getClass().getAnnotation(PluginDescriptor.class).name()))
 				.forEach(plugin ->
 				{
+					final PluginCategory pluginCategory = plugin.getClass()
```
Comment (deathbeam) [inline runelite-client/src/main/java/net/runelite/client/plugins/config/ConfigPanel.java]: Get the annotation only once and use it for both category and plugin name

**Comment by AdamBallinger** [comment]
531fb1ac7132823a29941a7a6e45d910
I have removed the dropdown box and instead display a label for each category with a list of plugins under it that are assigned to that category.
I have also removed the ALL category and with it the default category for PluginDescriptor.

**Comment by AdamBallinger** [comment]
As the obsidian theme was merged I have made changes to make the style of the category labels consistent with the rest of the theme. 
ae80ac83c5d7984e85d35676d8f5e73e
Let me know if it can be improved in any way.

**Pull request closed without merging**